[JSC] Add `"never"` and `"formalSymbol"` to `currencyDisplay` option for `Intl.NumberFormat`

sosukesuzuki opened the pull request:
#### 8e35dc0335a2dae0574990ba5469788ddbe86fd6
<pre>
[JSC] Add `&quot;never&quot;` and `&quot;formalSymbol&quot;` to `currencyDisplay` option for `Intl.NumberFormat`
<a href="https://bugs.webkit.org/show_bug.cgi?id=284613">https://bugs.webkit.org/show_bug.cgi?id=284613</a>

Reviewed by Yusuke Suzuki.

This patch adds support for More Currency Display Choices stage2
proposal[1]. This proposal adds `&quot;formalSymbol&quot;` and `&quot;never&quot;`
to `currencyDisplay` option for `Intl.NumberFormat`.

[1]: <a href="https://github.com/tc39/proposal-intl-currency-display-choices">https://github.com/tc39/proposal-intl-currency-display-choices</a>

* JSTests/stress/intl-numberformat-currencyDisplay-formalSymbol.js: Added.
(shouldBe):
(throw.new.Error):
* JSTests/stress/intl-numberformat-currencyDisplay-never.js: Added.
(shouldBe):
(throw.new.Error):
* Source/JavaScriptCore/runtime/IntlNumberFormat.cpp:
(JSC::IntlNumberFormat::initializeNumberFormat):
(JSC::IntlNumberFormat::currencyDisplayString):
* Source/JavaScriptCore/runtime/IntlNumberFormat.h:
* Source/JavaScriptCore/runtime/OptionsList.h:

Canonical link: <a href="https://commits.webkit.org/287849@main">https://commits.webkit.org/287849@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f3d643549d6fd367ea6f39b0259f5c4ab4b7617d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80841 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/363 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34777 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85368 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31825 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/380 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8162 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63125 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20905 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83910 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/189 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73577 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43429 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/105 "Build was cancelled. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Skipping applying patch since patch_id isn't provided; Checked out pull request") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27737 "Found 2 new test failures: imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/block-container-block-start-child-with-border.html imported/w3c/web-platform-tests/css/css-masking/clip-path/animations/clip-path-animation-font-size-mixed-change.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30282 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/73821 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71668 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28273 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86800 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/79900 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8068 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5692 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71421 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8245 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69413 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70662 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17637 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14691 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13626 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/102307 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8030 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24878 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/struct-new_default-small-members.js.wasm-collect-continuously (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7869 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11388 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9675 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on Source/JavaScriptCore/runtime/IntlNumberFormat.cpp:
```diff
@@ -419,12 +422,20 @@ void IntlNumberFormat::initializeNumberFormat(JSGlobalObject* globalObject, JSVa
             break;
+        case CurrencyDisplay::FormalSymbol:
+            ASSERT(Options::useMoreCurrencyDisplayChoices());
+            skeletonBuilder.append(" unit-width-formal"_s);
```
This is stable from ICU 68 ( https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/unumberformatter_8h.html#a5ee613b0f806f558d9ea877aa62a0c3f )

sosukesuzuki commented on Source/JavaScriptCore/runtime/IntlNumberFormat.cpp:
```diff
@@ -419,12 +422,20 @@ void IntlNumberFormat::initializeNumberFormat(JSGlobalObject* globalObject, JSVa
             break;
+        case CurrencyDisplay::Never:
+            ASSERT(Options::useMoreCurrencyDisplayChoices());
+            skeletonBuilder.append(" unit-width-hidden"_s);
```
This is stable from ICU 60

Constellation reviewed (approved):
r=me with title's typo `formalSynbol` => `formalSymbol`.

sosukesuzuki committed:
[JSC] Add `"never"` and `"formalSymbol"` to `currencyDisplay` option for `Intl.NumberFormat`
https://bugs.webkit.org/show_bug.cgi?id=284613

Reviewed by Yusuke Suzuki.

This patch adds support for More Currency Display Choices stage2
proposal[1]. This proposal adds `"formalSymbol"` and `"never"`
to `currencyDisplay` option for `Intl.NumberFormat`.

[1]: https://github.com/tc39/proposal-intl-currency-display-choices

* JSTests/stress/intl-numberformat-currencyDisplay-formalSymbol.js: Added.
(shouldBe):
(throw.new.Error):
* JSTests/stress/intl-numberformat-currencyDisplay-never.js: Added.
(shouldBe):
(throw.new.Error):
* Source/JavaScriptCore/runtime/IntlNumberFormat.cpp:
(JSC::IntlNumberFormat::initializeNumberFormat):
(JSC::IntlNumberFormat::currencyDisplayString):
* Source/JavaScriptCore/runtime/IntlNumberFormat.h:
* Source/JavaScriptCore/runtime/OptionsList.h:

Canonical link: https://commits.webkit.org/287849@main
